File: dream/双指针/No670.java
package 双指针;

/**
 * @author <a href="https://github.com/Matthew-Han">Matthew Han</a>
 * @date 2022/9/15 09:59 15
 * @since 1.0
 **/
public class No670 {
    /**
     * 最大交换
     *
     * @param num
     * @return
     */
    public int maximumSwap(int num) {
        char[] nums = String.valueOf(num).toCharArray();
        int l = 0;
        while (l < nums.length) {
            int r = nums.length - 1;
            int idx = r;
            char max = nums[r];
            while (r-- > l) {
                if (nums[r] > max) {
                    max = nums[r];
                    idx = r;
                }
            }
            if (max > nums[l]) {
                nums[l] ^= nums[idx];
                nums[idx] ^= nums[l];
                nums[l] ^= nums[idx];
                break;
            }
            l++;
        }
        return Integer.parseInt(new String(nums));
    }

}

File: lang-py/No670.py
"""
  author <a href="https://github.com/Matthew-Han">Matthew Han</a>
  since 2024/1/22 10:39:31
"""
from typing import List


class Solution:
    def maximumSwap(self, num: int) -> int:
        nums = list(list(str(num)))
        flag = False
        t1, t2 = 0, 0
        max = '0'
        for j in range(len(nums)):
            for i in range(len(nums) - 1, j, -1):
                if nums[i] > nums[j] and nums[i] > max:
                    t1 = i
                    t2 = j
                    max = nums[i]
                    flag = True
            if flag:
                break
        t = nums[t1]
        nums[t1] = nums[t2]
        nums[t2] = t
        return int(''.join(nums))
